ibis/bigquery/udf/udf.py

```diff
@@ -0,0 +1,820 @@
+"""Translate a Python AST to JavaScript
+"""
+
+import ast
+import collections
+import contextlib
+import functools
+import inspect
+import textwrap
+
+import six
+
+from multipledispatch import Dispatcher
+
+import ibis.expr.datatypes as dt
+
+from ibis.bigquery.udf.find import find
+from ibis.bigquery.udf.rewrite import rewrite
+from ibis.bigquery.compiler import compiles, BigQueryUDFNode
+
+
+ibis_type_to_bigquery_type = Dispatcher('ibis_type_to_bigquery_type')
+
+
+@ibis_type_to_bigquery_type.register(dt.Floating)
+def trans_float64(t):
+    return 'FLOAT64'
+
+
+@ibis_type_to_bigquery_type.register(dt.Integer)
+def trans_integer(t):
+    # BigQuery doesn't accept integer types because javascript doesn't
+    return 'FLOAT64'
+
+
+@ibis_type_to_bigquery_type.register(dt.Array)
+def trans_array(t):
+    return 'ARRAY<{}>'.format(ibis_type_to_bigquery_type(t.value_type))
+
+
+@ibis_type_to_bigquery_type.register(dt.Struct)
+def trans_struct(t):
+    return 'STRUCT<{}>'.format(
+        ', '.join(
+            '{} {}'.format(name, ibis_type_to_bigquery_type(type))
+            for name, type in zip(t.names, t.types)
+        )
+    )
+
+
+@ibis_type_to_bigquery_type.register(dt.Date)
+def trans_date(t):
+    return 'DATE'
+
+
+@ibis_type_to_bigquery_type.register(dt.Timestamp)
+def trans_timestamp(t):
+    if t.timezone is not None:
+        raise TypeError('BigQuery does not support timestamps with timezones')
+    return 'TIMESTAMP'
+
+
+@ibis_type_to_bigquery_type.register(dt.DataType)
+def trans_type(t):
+    return str(t).upper()
+
+
+@ibis_type_to_bigquery_type.register(six.string_types)
+def trans_str(t):
+    return ibis_type_to_bigquery_type(dt.dtype(t))
+
+
+class SymbolTable(collections.ChainMap):
+    """ChainMap subclass implementing scope for the translator.
+
+    Notes
+    -----
+    JavaScript requires declarations in strict mode, so to implement this we
+    shove a "let" at the beginning of every variable name if it doesn't already
+    exist in the current scope.
+    """
+    def __getitem__(self, key):
+        if key not in self:
+            self[key] = key
+            return 'let {}'.format(key)
+        return key
+
+
+def indent(lines, spaces=4):
+    """Indent `lines` by `spaces` spaces.
+
+    Parameters
+    ----------
+    lines : Union[str, List[str]]
+        A string or list of strings to indent
+    spaces : int
+        The number of spaces to indent `lines`
+
+    Returns
+    -------
+    indented_lines : str
+    """
+    if isinstance(lines, six.string_types):
+        text = [lines]
+    text = '\n'.join(lines)
+    return textwrap.indent(text, ' ' * spaces)
+
+
+def udf(input_type, output_type, strict=True):
```

Comment:
How about moving this to `ibis.bigquery.udf.api` or something?

Reply:
done